ARROW-16901: [R][CI] Prune R nightly builds

assignUser opened the pull request:
This PR adds pruning to the nightly R upload, 14 versions will be kept by default. 

I have removed the `burnett01/rsync-deployments` actions because the use of docker for this was unnecessary and the action can only upload to a remote. This new manual version also utilizes host key checking for which I created `secrets.NIGHTLIES_RSYNC_HOST_KEY` (which should contain the result of ` ssh-keyscan -H nightlies.apache.org 2> /dev/null` and needs to be added to apache/arrow before this can run).  This way we are no longer depending on the action and it's associated Dockerfile (`drinternet/rsync`). 

We might want to refactor this into a local action for use with all nightly upload jobs.

The pruning is not super efficient as we download the whole nightly repository (on cache miss). This could be avoided for the libarrow files, they could possibly be deleted via ssh instead but we need to download all R packages as `tools::write_PACKAGES` needs access to each archive.

assignUser commented:
@raulcd could you give this a first pass?

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -100,18 +149,37 @@ jobs:
+          # prune on each leaf dir and not on repo/.
+          for dir in ${repo_dirs[@]}; do
+            prune $dir
+          done
```
Can we use `find repo/* -mtime +$KEEP -type f -print0 | xargs -0 rm -rf` here?

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -59,6 +65,12 @@ jobs:
+          path: repo
+          key: r-nightly-${{ github.run_id }}
+          restore-keys: r-nightly-
+      
```
It seems this is duplicated.

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -124,14 +192,32 @@ jobs:
-          remote_user: ${{ secrets.NIGHTLIES_RSYNC_USER }}
-          remote_key: ${{ secrets.NIGHTLIES_RSYNC_KEY }}
+        run: tree repo
+      - name: Sync to Remote
```
Why do we implement this instead of using `burnett01/rsync-deployments@5.2`?

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +89,47 @@ jobs:
+          path: repo
+          key: r-nightly-${{ github.run_id }}
+          restore-keys: r-nightly-
+      - name: Sync from Remote
```
I think that we don't need this if we use `actions/cache` for `repo/`.
If we use `actions/cache` for `repo/`, `repo/` has the same contents of https://nightlies.apache.org/arrow/r/ eventually.

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +86,47 @@ jobs:
+
+            # Deploy. It is important to append '/*' to the source path otherwise
+            # the entire source dir will be created as a sub dir in the destination
+            sh -c "rsync $RSYNC_SWITCHES -e '$RSH' $DSN:$REMOTE_PATH/* $LOCAL_PATH"
```
Why do we need `sh -c`?

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +86,47 @@ jobs:
+            # Deploy. It is important to append '/*' to the source path otherwise
+            # the entire source dir will be created as a sub dir in the destination
+            sh -c "rsync $RSYNC_SWITCHES -e '$RSH' $DSN:$REMOTE_PATH/* $LOCAL_PATH"
+      - run: tree repo
```
Could you show an output of this?

I don't know why we don't use the same directory structure for `repo/` and https://nightlies.apache.org/arrow/r/ . If we use the same directory structure, we don't need to append `/*` for `rsync` source.

assignUser replied:
A small example:
```
$ tree src
src
└── arrow
    └── r
        └── pkg.file

2 directories, 1 file
$ tree dest
dest

0 directories, 0 files
$ rsync -avzh --update --delete src dest
sending incremental file list
src/
src/arrow/
src/arrow/r/
src/arrow/r/pkg.file

sent 180 bytes  received 51 bytes  462.00 bytes/sec
total size is 0  speedup is 0.00
$ tree dest
dest
└── src
    └── arrow
        └── r
            └── pkg.file

3 directories, 1 file
```
vs

```
$ rsync -avzh --update --delete src/* dest
sending incremental file list
arrow/
arrow/r/
arrow/r/pkg.file

sent 158 bytes  received 47 bytes  410.00 bytes/sec
total size is 0  speedup is 0.00
$ tree dest
dest
└── arrow
    └── r
        └── pkg.file

2 directories, 1 file
```

kou replied:
Thanks.
Could you try `rsync -avzh --update --delete src/ dest`? (`/` is appended to `src`.)

assignUser replied:
Yes that works!

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +86,47 @@ jobs:
+        run: |
+            # Make SSH key available and add remote to known hosts
+            eval "$(ssh-agent)" > /dev/null
+            echo "$SSH_KEY" | tr -d '\r' | DISPLAY=1 ssh-add - >/dev/null
```
Why do we need `DISPLAY=1`?
I think that there is no X on GitHub Actions.

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +86,47 @@ jobs:
+            # Make SSH key available and add remote to known hosts
+            eval "$(ssh-agent)" > /dev/null
+            echo "$SSH_KEY" | tr -d '\r' | DISPLAY=1 ssh-add - >/dev/null
+            mkdir -p ~/.ssh
```
Do we need `chmod go-rwx ~/.ssh`?

kou commented on .github/workflows/r_nightly.yml:
```diff
@@ -77,10 +86,47 @@ jobs:
+
+            # avzrh - archive mode, verbose, with compression, recurse into dirs, human readable numbers
+            # transfer only new files, delete files not in source, show progress info
+            RSYNC_SWITCHES="-avzrh --update --delete --progress"
```
`-r` is redundant because `a` includes `-r`:

```bash
$ rsync --help
....
       --archive, -a            archive mode is -rlptgoD (no -A,-X,-U,-N,-H)
...
```

assignUser commented:
Thanks for the review @kou ! While integrating the changes I was already annoyed to do everything twice so I refactored the syncing as a local action. This way it can be easily reused by other components to upload their nightly builds (e.g. java).
I also switched to using a local `.ssh` folder for `known_hosts` to avoid interfering with the runner state (not strictly necessary but I think good form).

Here is an example run in a test repo
https://github.com/assignUser/test-repo-a/runs/7130803918?check_suite_focus=true

kou commented on .github/actions/sync-nightlies/action.yml:
```diff
@@ -0,0 +1,92 @@
+    description: 'The remote key'
+    required: true
+  remote_host_key:
+    description: 'The host key fot Stricthostkeychecking'
```
```suggestion
    description: 'The host key for StrictHostKeyChecking'

```

kou commented on .github/actions/sync-nightlies/action.yml:
```diff
@@ -0,0 +1,92 @@
+      shell: bash
+      env:
+          SWITCHES: "${{ inputs.switches }}"
+          LOCAL_PATH: "${{ github.workspace}}/${{ inputs.local_path }}"
```
```suggestion
          LOCAL_PATH: "${{ github.workspace }}/${{ inputs.local_path }}"

```

raulcd commented on .github/workflows/r_nightly.yml:
```diff
@@ -101,18 +127,37 @@ jobs:
+      - name: Prune Repository
+        shell: bash
+        env:
+          KEEP: ${{ github.event.inputs.keep || 14 }}
```
shouldn't this already be the default value? I am not sure I understand why do we need the `|| 14`

assignUser commented on .github/actions/sync-nightlies/action.yml:
```diff
@@ -0,0 +1,92 @@
+        # We have to use a custom RSH to supply the port
+        RSH="ssh -o UserKnownHostsFile=.ssh/known_hosts -p $PORT"
+        DSN="$USER@$HOST"
+
+        # It is important to append '/*' to the source path otherwise
+        # the entire source dir will be created as a sub dir in the destination
+        if [ "${{ inputs.upload }}" = true ]
+        then
+          SOURCE=$LOCAL_PATH/*
+          DEST=$DSN:$REMOTE_PATH
+        else
+          SOURCE=$DSN:$REMOTE_PATH/*
+          DEST=$LOCAL_PATH
+        fi
```
```suggestion

        # It is important to append '/' to the source path otherwise
        # the entire source dir will be created as a sub dir in the destination
        if [ "${{ inputs.upload }}" = true ]
        then
          SOURCE=$LOCAL_PATH/
          DEST=$DSN:$REMOTE_PATH
        else
          SOURCE=$DSN:$REMOTE_PATH/
          DEST=$LOCAL_PATH
        fi

```

kou commented on .github/actions/sync-nightlies/action.yml:
```diff
@@ -0,0 +1,95 @@
+        RSH="ssh -o UserKnownHostsFile=.ssh/known_hosts -p $PORT"
+        DSN="$USER@$HOST"
+
+        # It is important to append '/' to the source path otherwise
```
How about appending `/` in caller not here to control how to synchronize by caller?

e.g.:

```yaml
        with:
          local_path: repo/
```

```yaml
        with:
          remote_path: ${{ secrets.NIGHTLIES_RSYNC_PATH }}/arrow/r/
```

assignUser replied:
Than the caller needs to know what happens within the action, which should not be required imho. I see your point about flexibility, this would break if I tried to sync just one file due to the appended `/`.

How about we check if `SOURCE` is a directory or file and only add a slash if it is a directory and does not already have a trailing slash?

kou replied:
> How about we check if `SOURCE` is a directory or file and only add a slash if it is a directory and does not already have a trailing slash?

If we use this approach, we can't copy `src/` to `dest/src/` by `local_path: src` (no trailing slash).

If you don't like the "appending `/` in caller` approach, I'm OK with the current approach. But could you add a comment that "this action supports only directory as source" or something?

kou commented on .github/actions/sync-nightlies/README.md:
```diff
@@ -0,0 +1,51 @@
+# Sync Nightlies
```
Could you add a license header?

kou reviewed (approved):
+1
